src/tools/compiletest/src/runtest.rs

```diff
@@ -709,8 +709,12 @@ impl<'test> TestCx<'test> {
             self.testpaths.file.display().to_string()
         };
 
-        let expect_help = expected_errors.iter().any(|ee| ee.kind == Some(ErrorKind::Help));
-        let expect_note = expected_errors.iter().any(|ee| ee.kind == Some(ErrorKind::Note));
+        // Errors and warnings are always expected, other diagnostics are only expected
+        // if one of them actually occurs in the test.
+        let expected_kinds: HashSet<_> = [ErrorKind::Error, ErrorKind::Warning]
+            .into_iter()
+            .chain(expected_errors.iter().filter_map(|e| e.kind))
+            .collect();
```

Comment:
Suggestion: could you document this behavior in rustc-dev-guide? And also document `dont-require-annotations` (I forgot to ask in https://github.com/rust-lang/rust/pull/139489)?

Reply:
Ah, right. Will do.

Reply:
Added the docs.